util/linuxfw/nftables_runner.go

```diff
@@ -0,0 +1,1068 @@
+	}
+	if strings.HasPrefix(hexStr, "0x") {
+		hexStr = strings.TrimPrefix(hexStr, "0x")
+	}
+
+	temp, err := hex.DecodeString(hexStr)
+	if err != nil {
+		return nil, fmt.Errorf("decode hex string: %w", err)
+	}
+
+	ret := make([]byte, toLen)
+	copy(ret[toLen-len(temp):], temp)
+	return ret, nil
+}
+
+// createTableIfNotExist creates a nftables table via connection c if it does not exist within the given family.
+func createTableIfNotExist(c *nftables.Conn, family nftables.TableFamily, name string) (*nftables.Table, error) {
+	tables, err := c.ListTables()
+	if err != nil {
+		return nil, fmt.Errorf("get tables: %w", err)
+	}
+	for _, table := range tables {
+		if table.Name == name && table.Family == family {
+			return table, nil
+		}
+	}
+
+	t := c.AddTable(&nftables.Table{
+		Family: family,
+		Name:   name,
+	})
+	if err := c.Flush(); err != nil {
+		return nil, fmt.Errorf("add table: %w", err)
+	}
+	return t, nil
+}
+
+type errorChainNotFound struct {
+	chainName string
+	tableName string
+}
+
+func (e *errorChainNotFound) Error() string {
+	return fmt.Sprintf("chain %s not found in table %s", e.chainName, e.tableName)
+}
+
+// getChainFromTable returns the chain with the given name from the given table.
+// Note that chain name is unique within a table.
+func getChainFromTable(c *nftables.Conn, table *nftables.Table, name string) (*nftables.Chain, error) {
+	chains, err := c.ListChainsOfTableFamily(table.Family)
+	if err != nil {
+		return nil, fmt.Errorf("list chains: %w", err)
+	}
+
+	for _, chain := range chains {
+		// Table family is already checked so table name is unique
+		if chain.Table.Name == table.Name && chain.Name == name {
+			return chain, nil
+		}
+	}
+
+	return nil, &errorChainNotFound{table.Name, name}
+}
+
+// getChainsFromTable returns all chains from the given table.
+func getChainsFromTable(c *nftables.Conn, table *nftables.Table) ([]*nftables.Chain, error) {
+	chains, err := c.ListChainsOfTableFamily(table.Family)
+	if err != nil {
+		return nil, fmt.Errorf("list chains: %w", err)
+	}
+
+	var ret []*nftables.Chain
+	for _, chain := range chains {
+		// Table family is already checked so table name is unique
+		if chain.Table.Name == table.Name {
+			ret = append(ret, chain)
+		}
+	}
+
+	return ret, nil
+}
+
+// createChainIfNotExist creates a chain with the given name in the given table
+// if it does not exist.
+func createChainIfNotExist(
+	c *nftables.Conn, table *nftables.Table,
+	name string, chainType nftables.ChainType,
+	chainHook *nftables.ChainHook,
+	chainPriority *nftables.ChainPriority) error {
+	_, err := getChainFromTable(c, table, name)
+	if err != nil {
+		_, ok := err.(*errorChainNotFound)
+		if !ok {
+			return fmt.Errorf("get chain: %w", err)
+		}
+	}
+
+	_ = c.AddChain(&nftables.Chain{
+		Name:     name,
+		Table:    table,
+		Type:     chainType,
+		Hooknum:  chainHook,
+		Priority: chainPriority,
+	})
+
+	if err := c.Flush(); err != nil {
+		return fmt.Errorf("add chain: %w", err)
```

Comment:
This Flush is flushing the add chain operation so I logged as add chain. I did this because AddChain doesn't return right away so the only way we know if it's correctly run is by flushing right after.